### PR TITLE
Obc 349 add ability to rename metrics

### PR DESCRIPTION
**Pull request by Tenaria**
Adds in feature to enable ability to rename metrics based off attributes as discussed with @chenzhihao 

For implementation details, see https://hello.atlassian.net/wiki/spaces/~315498988/pages/1498526645/span+metrics+processor+rename+metrics+feature

### CODE REVIEW

**Comment by jamesmoessis** [comment]
This seems like the wrong place to be renaming metrics. If possible perhaps you should utilise the `metricstransformprocessor`?

**Comment by Tenaria** [comment]
@jamesmoessis 👋 

imo it makes sense as a part of the span metrics processor for a few reasons:
* metrics generated should conform to the semantic convention. This enables a user to apply this. imo it should be a core feature as it's a bit weird for a OTel processor to be generating something that doesn't conform to the conventions.
* It also allows renaming the default metric names `calls_total` and `latency` which should probably be up to the user, there's no defined convention for this.
* It's less computationally expensive if we do it natively as a part of the processor directly on metric creation instead of requiring to pass all metrics through another processor to be checked against.
* Following on from the above point, it also means that we can keep this naming requirement local to the span metrics processor. Instead of having all metrics that go via the collector parsed (computationally expensive) we can only have the custom ones generated from the span metrics processor parsed and renamed. The alternative is to set up another receiver / collecter pipeline and have the span metrics processor recursively send back to the collector which I think can start to make it a bit messy and it would be good if we could solve these problems native in the span metrics processor.
* The logic is a config option that can be enabled and it doesn't particularly complicate the main logic too much (unlike aggregation which was not configurable and complicated the code whilst not being necessary since the OTel spec defines aggregation as a part of the client or backend and not the collector. On the other hand, the semantic convention is a defined piece and the generator/instrumenter/creator/library of the metrics should natively be able to handle this imo, not rely on the user to use a third-party processor to manipulate)

As an aside, the metrics transform processor is not in a usable state. It's using OpenCensus and doesn't have the full compatibilities so it's missing a few things that are quite important like instrumentation library and status codes. It'll end up removing this from all metrics in the pipeline as a result.

**Comment by jamesmoessis** [comment]
> metrics generated should conform to the semantic convention. This enables a user to apply this. imo it should be a core feature as it's a bit weird for a OTel processor to be generating something that doesn't conform to the conventions.

I agree it's weird that the processor doesn't really follow the convention. Going forward, that's something we should iterate over and work on. However, I don't really follow the point that it enables the user to apply the semantic convention. The metricstransformprocessor also allows this. 

> It also allows renaming the default metric names calls_total and latency which should probably be up to the user, there's no defined convention for this.

Again, this is one of the main features of the `metricstransformprocessor`. 

> Following on from the above point, it also means that we can keep this naming requirement local to the span metrics processor. Instead of having all metrics that go via the collector parsed (computationally expensive) we can only have the custom ones generated from the span metrics processor parsed and renamed.

You can do this in a non-messy way. The metrics right now get sent straight to an exporter (they "cross over" from spans pipeline to metrics pipeline). You would just replace that exporter with the `metricstransformprocessor` that's configured. So, only the generated metrics would pass through that processor. 

All this being said, if the metrics transform processor isn't usable, then you can't do much. I still however think we are taking a very monolithic approach to this processor and it might hurt us when trying to iterate and upgrade how we do span-metrics conversion. Ideally the end state doesn't have anything custom

**Comment by Tenaria** [comment]
> I agree it's weird that the processor doesn't really follow the convention. Going forward, that's something we should iterate over and work on. However, I don't really follow the point that it enables the user to apply the semantic convention. The metricstransformprocessor also allows this.

I chose to do it as configuration because I wanted to allow the flexibility of how users want to use it rather than hard coding and forcing a convention (which btw the convention is very flexible given that it has a whole bunch of optional things and etc.). It also allows the users who understand their application to explicitly define what they want instead of having to check through every piece of the semantic convention which can be expensive and for those with a bit of custom convention :badpokerface: their own custom renaming. We can think about having it hardcoded in the future, but for now I think keeping it flexible is more ideal.

> Again, this is one of the main features of the metricstransformprocessor.

It is a feature but I feel like it should be a core configuration on the span metrics processor. It doesn't really make sense to build metrics and then have to check every metric if it needs to be renamed separately. It's also more computationally efficient because it directly changes the name rather then building it then iterating over every metric that comes through to check if we need to rename.

> You can do this in a non-messy way. The metrics right now get sent straight to an exporter (they "cross over" from spans pipeline to metrics pipeline). You would just replace that exporter with the metricstransformprocessor that's configured. So, only the generated metrics would pass through that processor.

I'm not sure I quite follow what you mean here if you could elaborate. A processor is a piece of functionality that does some logic over metrics/traces it receivers from a receiver. The processor then passes all its data to an exporter once it's done. The span metrics processor is a little different in that it acts as a processor on a span pipeline then **directly** exports the metric to an exporter. That means that if you want a processor to do something to the metrics, you need to set the exporter on the span metrics processor to send back to the receiver from the collector. Which then it'll pass on to the processor and then export to whatever backend you have set.

> All this being said, if the metrics transform processor isn't usable, then you can't do much. I still however think we are taking a very monolithic approach to this processor and it might hurt us when trying to iterate and upgrade how we do span-metrics conversion. Ideally the end state doesn't have anything custom

The metrics transform processor isn't usable 😅 I don't necessarily think it's monolithic since it should be a core feature of the span metrics processor. If you want to be pure monolithic then I would argue that specifying what attributes you want to be pulled out of the span metrics processor is an anti-pattern and that we should pull out all attributes and all resource attributes in the span metrics processor then rely on the metrics transform processor to remove the attributes you don't want. That would be pure monolothic and very inefficient. But you need to draw the line somewhere and decide what needs to be core functionality of the span metrics processor in a "monolithic" fashion and what should be generalised functionality. In this case I think that renaming to allow for conforming to the semantic convention should be core functionality.

**Comment by jamesmoessis** [comment]
@Tenaria ok, I see your points. By the way, which semantic convention would you be using? The one we have defined internally or the OTel metrics semconv? They are similar but have some differences. I tend to lean towards using OTel's semconv. The problem is, it isn't stable.

**Comment by Tenaria** [comment]
> By the way, which semantic convention would you be using
Probably not the answer you wanted to hear, but I would like to say our convention, since that's what we / our users should be using. 

If we want to use the upstream convention, we should update our convention and then all our internal tools should follow suit since they are following our convention. That way everything is using the same convention and we have a single source of truth, instead of getting confused and having the span metrics processor use the upstream convention and then the java client library use our Semantic Convention and then something else later use the upstream convention which has now had its version updated.

**Comment by chenzhihao** [comment]
I think this surely could be done in `metricstransformprocessor` if that's usable. But it also makes sense to do it here. As we don't have a convention for these R.E.D. metrics(please correct me if I'm wrong), `calls_total` and `latency` seems arbitrary. It makes sense to have the ability to configure the metrics name here.

As this fork will eventually be ditched, it's fairly practical to add this feature here to meet our requirements.

File: processor/spanmetricsprocessor/config.go [hunk 1]
```diff
@@ -86,6 +86,39 @@ type Config struct {
 	// LaunchDarklyKey defines the LaunchDarkly key.
 	// Optional. Only required when `EnableFeatureFlag` is `true`.
 	LaunchDarklyKey string `mapstructure:"launch_darkly_key"`
+
+	// Transforms takes in a list of Actions to perform on metrics generated by span metrics processor.
+	// currently there is only one action, the rename action. This allows a user to specify metrics with certain attributes
+	// to have a specific metric name. By default, the generated metric names for Rate / Error is "calls_total" and Duration is
+	// "latency".
+	Transforms []Transform `mapstructure:"transforms"`
```
Comment (chenzhihao) [inline processor/spanmetricsprocessor/config.go:94]: How about just calling this `Renames []Rename`? The struct itself is dedicated to renaming transform already.

File: processor/spanmetricsprocessor/config.go [hunk 2]
```diff
@@ -86,6 +86,39 @@ type Config struct {
...
+// If no attributes are specified, then all metrics will be caught under this case and renamed.
+type Transform struct {
+	Attributes              []Dimension `mapstructure:"attributes"`
+	NewCallsTotalMetricName string      `mapstructure:"new_calls_total_metric_name"`
+	NewLatencyMetricName    string      `mapstructure:"new_latency_metric_name"`
+}
+
+func (t Transform) attributesMatched(attributesOnMetric *pdata.AttributeMap) bool {
+	// If no attribute specified then it is the default/ catch-all case
+	if len(t.Attributes) == 0 {
```
Comment (chenzhihao) [inline processor/spanmetricsprocessor/config.go:108]: change to `allAttributesMatched` will eliminate the ambiguity.

File: processor/spanmetricsprocessor/config.go [hunk 3]
```diff
@@ -86,6 +86,39 @@ type Config struct {
...
 	LaunchDarklyKey string `mapstructure:"launch_darkly_key"`
+
+	// Transforms takes in a list of Actions to perform on metrics generated by span metrics processor.
+	// currently there is only one action, the rename action. This allows a user to specify metrics with certain attributes
+	// to have a specific metric name. By default, the generated metric names for Rate / Error is "calls_total" and Duration is
+	// "latency".
+	Transforms []Transform `mapstructure:"transforms"`
+}
+
+// Transform defines the new metric names and the attributes to match on in order for the rename to occur
```
Comment (chenzhihao) [inline processor/spanmetricsprocessor/config.go:97]: I'm confused about the renaming behaviour.

I think that should be a `Rule` type as `matchAny`, `matchAll`, `matchNone` to cover the cases. But we can start with what we need. Looks like we want a `matchAll` rule for our requirements?

File: processor/spanmetricsprocessor/processor.go [hunk 4]
```diff
@@ -51,6 +51,9 @@ const (
 
 	defaultDimensionsCacheSize         = 1000
 	defaultResourceAttributesCacheSize = 1000
+
+	defaultCallsTotalMetricName = "calls_total"
+	defaultLatencyMetricName    = "latency"
```
Comment (chenzhihao) [inline processor/spanmetricsprocessor/processor.go:56]: if we create this processor starts from scratch, will we still call it `latency`? As the metrics are not only HTTP metrics, `duration` probably makes more sense.
As it's already named as this from the upstream, we can just leave it as it is.

File: processor/spanmetricsprocessor/processor_test.go [hunk 5]
```diff
@@ -376,6 +378,37 @@ func TestProcessorConsumeTraces(t *testing.T) {
 			},
 			traces: []pdata.Traces{spanWithLargeTimestamp},
 		},
+		{
+			name:                   "Test that metric renaming works",
```
Comment (chenzhihao) [inline processor/spanmetricsprocessor/processor_test.go:382]: Should we have another case to cover the default metric names `calls_total` and `latency`?

File: processor/spanmetricsprocessor/processor.go (hunk 4 above)
Comment (Tenaria) [reply processor/spanmetricsprocessor/processor.go:56]: I think `latency` can still be ok. You can still have database latency (time it takes to get back to you), network latency, etc.. I honestly don't feel strongly about it though, I think `latency` and `duration` can both work.

Comment (jamesmoessis) [reply processor/spanmetricsprocessor/processor.go:56]: The ideal end state is a mapping from the trace convention to the metrics convention. e.g. `http.client.duration -> metric.http.client.duration`. Not asking for that here, but that's what we should be looking to eventually do.

**Comment by jamesmoessis** [comment]
> If we want to use the upstream convention, we should update our convention and then all our internal tools should follow suit since they are following our convention.

Once the metrics semantic convention (in OTel) gains more stability this is what we definitely should do. I don't think we should be defining our own semconvs at all. All the useful instrumentation out there follows OTel, not our similarly defined metrics semconv.

File: processor/spanmetricsprocessor/processor.go (hunk 4 above)
Comment (Tenaria) [reply processor/spanmetricsprocessor/processor.go:56]: I think this default name should mostly for custom traces that don't follow a convention. Although ideally we really shouldn't be doing anything custom but I think that'll be a hard ask 😅 somewhere along the way someone will probably want something custom and have some valid reason for it

**Review by chenzhihao — APPROVED** [review]

**Pull request merged**